docs/architecture-decisions/2024-05-22_teal-compilation.md

```diff
@@ -0,0 +1,84 @@
+# Compilation of Algorand TEAL Programs into AVM Bytecode
+
+- **Status**: Proposed
+- **Owner**: Daniel McGregor (MakerX)
+- **Deciders**: Rob Moore (MakerX), Adam Chidlow (MakerX), Alessandro Cappellato (Algorand Foundation)
+- **Date created**: 2024-05-22
+- **Date decided**: 
+- **Date updated**: 
+
+## Context
+
+The Puya compiler needs to assemble valid Algorand TEAL (Transaction Execution Approval Language) programs into AVM (Algorand Virtual Machine) bytecode 
+so the bytecode can be used within inner transactions that need to create applications.
+
+There are several potential solutions to achieve this compilation, each with its own advantages and trade-offs.
+
+## Requirements
+
+* The solution works cross-platform (i.e. works on Windows, Mac and Linux)
+* The solution is fast
+* The solution is correct
+
+## Options
+
+### Option 1. Use Algod Compile Endpoint
+
+Utilize the Algorand algod `/v2/teal/compile` endpoint to compile TEAL programs into AVM bytecode.
+
+**Pros**
+
+* Ensures compatibility with the latest Algorand protocol updates.
+* Existing solution
+
+**Cons**
+
+* Requires network connectivity to the Algorand API
```

Comment:
Not just network connectivity, but also configuration to a node, which adds complexity to how you invoke the compiler (it could obviously have some defaults, but given it causes network calls that may not be expected behaviour). Also, a compiler that needs out of process communication to compile is pretty unusual architecturally.